accepted/2024/trusted-publishers-oidc-for-nuget-push.md

```diff
@@ -0,0 +1,538 @@
+
+![Trust policy form](meta/trusted-publishers-oidc-for-nuget-push/trust-policy-form.png)
+
+In the form, the package author would provide 3 critical pieces of information:
+
+1. The GitHub organization or username owning the repository.
+2. The GitHub repository name.
+3. Some filter on which GitHub Action workflow runs should be allowed to push packages.
+
+The filter on which workflow runs should be allowed to publish packages can potentially be very specific. GitHub Actions
+supplies a variety of metadata in the OIDC token available to workflow runs. Refer to [Understanding the OIDC
+token](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#understanding-the-oidc-token)
+to see all of the properties.
+
+For the sake of simplicity, we will start with the following workflow filters:
+- Filtering for a specific *branch*, as shown in the UI above (branch filter)
+- Filtering for a specific *environment* (a property optionally set in the workflow)
+- Filtering for a specific *workflow path* (relative file path within the repository)
+
+The screenshot above only shows the branch filter, so the UI will need to be expanded to cover the desired workflow
+filters. A suggest UI would have 3 checkboxes:
+- [ ] Filter by branch: __________
+- [ ] Filter by environment: __________
+- [ ] Filter by workflow: __________
+
+At least one filter would need to be checked as to prevent an overly broad trust policy.
+
+Additional filters can be added in the future based on other properties provided in the GitHub OIDC token.
+
+In short, by adding this trust policy, the package author is saying:
+
+> I trust GitHub to only issue tokens for my repository for my GitHub Actions workflow *and* these tokens are sufficient
+> credentials for uploading packages to my NuGet.org account.
+
+Once the package author has added this trust policy, NuGet.org will trust GitHub Actions OIDC tokens matching this
+pattern much like an existing long-lived API key, with the added benefit of the trust policy never expiring.
+
+NuGet package sources must ensure a secure, authenticated session for the user that is adding the trust policy.
+NuGet.org requires two-factor authentication for all web UI sign-ins.
+
+#### What NuGet API operations can be used inside the GitHub Actions workflow?
+
+To scope the ways in which a GitHub trust policy can be used, similar scoping rules to the current API key flow can be
+set on the GitHub Actions trust policy:
+
+![Scope form](meta/trusted-publishers-oidc-for-nuget-push/scope-form.png)
+
+Just like the existing API scoping feature, there are 3 dimensions which a trust policy can be scoped. For more
+information about scopes on API keys, see [Scoped API
+keys](https://learn.microsoft.com/en-us/nuget/nuget-org/scoped-api-keys). In summary:
+
+- **Package owner**: for users that are members of one or more organizations, it's important to know which package owner
```

Comment:
Suppose you transfer ownership of a package to me.  (You make me a coowner, and then I remove you.)  How would I remove your ability to continue publishing packages with your trusted publisher policy?

Reply:
The moment you are removed as an owner, your trust policy will no longer work. The trust policy is moreso associated with a package owner rather than a package. In fact, the only this package-specific about a trust policy (or long lived API key) is a package ID scope.

Each time you upload a package version (via web, API key, or Trusted Publisher), whether or not the **Package owner** in question (the package owner property listed here) is a current, direct owner of the package. If it is not, the upload will be blocked.

I updated the spec to clarify this.